File: engine/normalizer.js
/**
 * STEP 3 — Normalizer (Linear Form)
 *
 * Converts a parsed equation AST into a standard linear form:
 *   a1*x1 + a2*x2 + ... = constant
 *
 * Output format:
 * {
 *   coefficients: { x: 10, y: 20 },
 *   constant: 100
 * }
 *
 * IMPORTANT: This normalizer intentionally supports LINEAR expressions only.
 * - Allowed: +, -, unary -, parentheses, implicit multiplication, number*expr
 * - Allowed: division by a non-zero integer constant if it divides evenly
 * - Rejected: variable * variable, division by variable/expression, non-integer results
 */

/** @typedef {{ type: 'NumberLiteral', value: number } | { type: 'Variable', name: string } | { type: 'UnaryExpression', op: 'u-', arg: AstNode } | { type: 'BinaryExpression', op: '+'|'-'|'*'|'/', left: AstNode, right: AstNode } AstNode */
/** @typedef {{ type: 'Equation', left: AstNode, right: AstNode }} EquationAst */

/** @typedef {{ constant: number, coefficients: Record<string, number> }} LinearForm */

function isSafeInt(n) {
  return Number.isSafeInteger(n);
}

function cloneCoeffs(c) {
  return { ...c };
}

function addCoeff(coeffs, variable, delta) {
  const next = (coeffs[variable] ?? 0) + delta;
  if (next === 0) {
    delete coeffs[variable];
  } else {
    coeffs[variable] = next;
  }
}

function normalizeCoeffs(coeffs) {
  for (const [k, v] of Object.entries(coeffs)) {
    if (v === 0) delete coeffs[k];
  }
  return coeffs;
}

function linearZero() {
  return { constant: 0, coefficients: {} };
}

function linearFromNumber(value) {
  if (!isSafeInt(value)) throw new Error("Only safe integers are supported.");
  return { constant: value, coefficients: {} };
}

function linearFromVariable(name) {
  return { constant: 0, coefficients: { [name]: 1 } };
}

function linearNeg(a) {
  const coeffs = {};
  for (const [k, v] of Object.entries(a.coefficients)) coeffs[k] = -v;
  return { constant: -a.constant, coefficients: coeffs };
}

function linearAdd(a, b) {
  const coeffs = cloneCoeffs(a.coefficients);
  for (const [k, v] of Object.entries(b.coefficients)) addCoeff(coeffs, k, v);
  const constant = a.constant + b.constant;
  if (!isSafeInt(constant)) throw new Error("Integer overflow in constant.");
  return { constant, coefficients: coeffs };
}

function linearSub(a, b) {
  return linearAdd(a, linearNeg(b));
}

function isConstantForm(f) {
  return Object.keys(f.coefficients).length === 0;
}

function linearScale(a, scalar) {
  if (!isSafeInt(scalar)) throw new Error("Scalar must be a safe integer.");
  const coeffs = {};
  for (const [k, v] of Object.entries(a.coefficients)) {
    const nv = v * scalar;
    if (!isSafeInt(nv)) throw new Error("Integer overflow in coefficient.");
    if (nv !== 0) coeffs[k] = nv;
  }
  const constant = a.constant * scalar;
  if (!isSafeInt(constant)) throw new Error("Integer overflow in constant.");
  return { constant, coefficients: coeffs };
}

function linearMul(a, b) {
  // Linear multiplication allowed only if one side is constant.
  if (isConstantForm(a)) return linearScale(b, a.constant);
  if (isConstantForm(b)) return linearScale(a, b.constant);
  throw new Error("Non-linear term detected: variable * variable.");
}

function linearDiv(a, denom) {
  // Linear division allowed only by a non-zero constant that divides evenly.
  if (!isConstantForm(denom)) {
    throw new Error("Division by a variable/expression is not supported.");
  }

  const d = denom.constant;
  if (d === 0) throw new Error("Division by zero.");

  if (a.constant % d !== 0) {
    throw new Error("Division produces non-integer constant.");
  }

  const coeffs = {};
  for (const [k, v] of Object.entries(a.coefficients)) {
    if (v % d !== 0) {
      throw new Error(`Division produces non-integer coefficient for ${k}.`);
    }
    coeffs[k] = v / d;
  }

  return { constant: a.constant / d, coefficients: normalizeCoeffs(coeffs) };
}

/**
 * Convert an AST expression to a linear form.
 * @param {AstNode} node
 * @returns {LinearForm}
 */
function astToLinearForm(node) {
  if (!node || typeof node !== "object") throw new Error("Invalid AST node.");

  if (node.type === "NumberLiteral") {
    return linearFromNumber(node.value);
  }

  if (node.type === "Variable") {
    return linearFromVariable(node.name);
  }

  if (node.type === "UnaryExpression") {
    if (node.op !== "u-") throw new Error(`Unsupported unary operator '${node.op}'.`);
    return linearNeg(astToLinearForm(node.arg));
  }

  if (node.type === "BinaryExpression") {
    const left = astToLinearForm(node.left);
    const right = astToLinearForm(node.right);

    if (node.op === "+") return linearAdd(left, right);
    if (node.op === "-") return linearSub(left, right);
    if (node.op === "*") return linearMul(left, right);
    if (node.op === "/") return linearDiv(left, right);

    throw new Error(`Unsupported binary operator '${node.op}'.`);
  }

  throw new Error(`Unknown AST node type '${node.type}'.`);
}

/**
 * Normalize a full equation AST into coefficients + constant.
 *
 * If left = right, and:
 *  left = Lc + sum(Li*xi)
 *  right = Rc + sum(Ri*xi)
 *
 * Then:
 *   sum((Li-Ri)*xi) = (Rc - Lc)
 *
 * @param {EquationAst} equation
 */
function normalizeEquation(equation) {
  if (!equation || equation.type !== "Equation") {
    throw new Error("Expected an Equation AST.");
  }

  const left = astToLinearForm(equation.left);
  const right = astToLinearForm(equation.right);

  /** @type {Record<string, number>} */
  const coefficients = {};

  const vars = new Set([...Object.keys(left.coefficients), ...Object.keys(right.coefficients)]);
  for (const v of vars) {
    const coeff = (left.coefficients[v] ?? 0) - (right.coefficients[v] ?? 0);
    if (coeff !== 0) coefficients[v] = coeff;
  }

  const constant = right.constant - left.constant;
  if (!isSafeInt(constant)) throw new Error("Integer overflow in constant.");

  return {
    coefficients,
    constant,
  };
}

module.exports = {
  astToLinearForm,
  normalizeEquation,
  linearZero,
};

File: engine/rust/src/normalizer.rs
// ============================================================
//  normalizer.rs — Port of engine/normalizer.js
//
//  Converts a parsed equation AST into standard linear form:
//    a1*x1 + a2*x2 + ... = constant
//
//  Only linear operations are supported:
//    +, -, unary -, parentheses, scalar*expr, expr/scalar_const
// ============================================================

use std::collections::HashMap;
use crate::parser::{AstNode, EquationAst};

/// Linear form: sum(coeff[v] * v) = constant
#[derive(Debug, Clone)]
pub struct LinearForm {
    pub coefficients: HashMap<char, i64>,
    pub constant: i64,
}

impl LinearForm {
    fn zero() -> Self {
        LinearForm { coefficients: HashMap::new(), constant: 0 }
    }

    fn from_number(n: i64) -> Self {
        LinearForm { coefficients: HashMap::new(), constant: n }
    }

    fn from_variable(v: char) -> Self {
        let mut c = HashMap::new();
        c.insert(v, 1);
        LinearForm { coefficients: c, constant: 0 }
    }

    fn is_constant(&self) -> bool {
        self.coefficients.is_empty()
    }

    fn negate(self) -> Self {
        LinearForm {
            coefficients: self.coefficients.into_iter().map(|(k, v)| (k, -v)).collect(),
            constant: -self.constant,
        }
    }

    fn add(mut self, other: LinearForm) -> Result<LinearForm, String> {
        let constant = self.constant.checked_add(other.constant)
            .ok_or("Integer overflow in constant.")?;
        for (k, v) in other.coefficients {
            let entry = self.coefficients.entry(k).or_insert(0);
            *entry = entry.checked_add(v).ok_or("Integer overflow in coefficient.")?;
            if *entry == 0 {
                self.coefficients.remove(&k);
            }
        }
        Ok(LinearForm { coefficients: self.coefficients, constant })
    }

    fn sub(self, other: LinearForm) -> Result<LinearForm, String> {
        self.add(other.negate())
    }

    fn scale(self, scalar: i64) -> Result<LinearForm, String> {
        let constant = self.constant.checked_mul(scalar)
            .ok_or("Integer overflow in constant.")?;
        let mut coefficients = HashMap::new();
        for (k, v) in self.coefficients {
            let nv = v.checked_mul(scalar)
                .ok_or("Integer overflow in coefficient.")?;
            if nv != 0 { coefficients.insert(k, nv); }
        }
        Ok(LinearForm { coefficients, constant })
    }

    fn mul(self, other: LinearForm) -> Result<LinearForm, String> {
        if self.is_constant() { return other.scale(self.constant); }
        if other.is_constant() { return self.scale(other.constant); }
        Err("Non-linear term: variable × variable.".to_string())
    }

    fn div(self, denom: LinearForm) -> Result<LinearForm, String> {
        if !denom.is_constant() {
            return Err("Division by a variable/expression is not supported.".to_string());
        }
        let d = denom.constant;
        if d == 0 { return Err("Division by zero.".to_string()); }

        if self.constant % d != 0 {
            return Err("Division produces non-integer constant.".to_string());
        }

        let mut coefficients = HashMap::new();
        for (k, v) in &self.coefficients {
            if v % d != 0 {
                return Err(format!("Division produces non-integer coefficient for {}.", k));
            }
            coefficients.insert(*k, v / d);
        }
        Ok(LinearForm { coefficients, constant: self.constant / d })
    }
}

fn ast_to_linear(node: &AstNode) -> Result<LinearForm, String> {
    match node {
        AstNode::NumberLiteral(n) => Ok(LinearForm::from_number(*n)),
        AstNode::Variable(c)     => Ok(LinearForm::from_variable(*c)),

        AstNode::UnaryExpr { arg } => {
            Ok(ast_to_linear(arg)?.negate())
        }

        AstNode::BinaryExpr { op, left, right } => {
            let l = ast_to_linear(left)?;
            let r = ast_to_linear(right)?;
            match op {
                '+' => l.add(r),
                '-' => l.sub(r),
                '*' => l.mul(r),
                '/' => l.div(r),
                _   => Err(format!("Unsupported binary operator '{}'.", op)),
            }
        }
    }
}

/// Normalize a full equation AST.
///
/// Converts left = right into:
///   (left_coeffs - right_coeffs) * vars = right_const - left_const
pub fn normalize_equation(eq: &EquationAst) -> Result<LinearForm, String> {
    let left  = ast_to_linear(&eq.left)?;
    let right = ast_to_linear(&eq.right)?;

    // Move all variable terms to left, all constants to right:
    //   left_v - right_v = right_c - left_c
    let mut coefficients: HashMap<char, i64> = HashMap::new();

    for (k, v) in &left.coefficients {
        *coefficients.entry(*k).or_insert(0) += v;
    }
    for (k, v) in &right.coefficients {
        *coefficients.entry(*k).or_insert(0) -= v;
    }
    // Remove zeros
    coefficients.retain(|_, v| *v != 0);

    let constant = right.constant.checked_sub(left.constant)
        .ok_or("Integer overflow in constant.")?;

    Ok(LinearForm { coefficients, constant })
}
